rs/ic_os/manifest_tool/src/check_unused_components.rs

```diff
@@ -0,0 +1,180 @@
+use anyhow::Result;
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+use walkdir::WalkDir;
+
+use crate::components_parser::{get_icos_manifest, IcosManifest, COMPONENTS_PATH};
+
+const IGNORED_COMPONENT_FILES: &[&str] = &[
+    "components/networking/dev-certs/root_cert_gen.sh",
+    "components/networking/dev-certs/canister_http_test_ca.key",
+];
+
+pub fn check_unused_components(repo_root: &Path) -> Result<()> {
+    let icos_manifest = get_icos_manifest(repo_root)?;
+
+    let components_path: PathBuf = repo_root.join(COMPONENTS_PATH);
+
+    let repo_files: HashSet<PathBuf> = collect_repo_files(&components_path)?;
+    let filtered_repo_files = filter_ignored_files(repo_files);
+
+    let manifest_files: HashSet<PathBuf> = collect_manifest_files(&icos_manifest);
+
+    let unused_files: Vec<&PathBuf> = filtered_repo_files.difference(&manifest_files).collect();
+
+    if unused_files.is_empty() {
+        println!("No unused files found.");
+        Ok(())
+    } else {
+        Err(anyhow::anyhow!(
+            "Unused files found:\n{}",
+            unused_files
+                .iter()
+                .map(|unused_file| unused_file.display().to_string())
+                .collect::<Vec<_>>()
+                .join("\n")
+        ))
+    }
+}
+
+fn collect_repo_files(dir: &Path) -> Result<HashSet<PathBuf>, std::io::Error> {
```

Comment:
As far as I can tell this will list all files in the repo, including potentially `.gitignore`d files (and might recurse down big cache directories like `target/`). Wouldn't `git ls-files` work better here?